server/src/main/java/org/opensearch/extensions/ExtensionsManager.java

```diff
@@ -522,15 +525,16 @@ public String executor() {
                 new IndicesModuleRequest(indexModule),
                 indicesModuleResponseHandler
             );
-            // TODO: make asynchronous
-            inProgressFuture.get(EXTENSION_REQUEST_WAIT_TIMEOUT, TimeUnit.SECONDS);
-            logger.info("Received response from Extension");
+            inProgressFuture.orTimeout(EXTENSION_REQUEST_WAIT_TIMEOUT, TimeUnit.SECONDS);
```

Comment:
I think There is a clear link between `inProgressFuture` and `inProgressIndexNameFuture` which could be composed using `whenComplete` callbacks:

```
private void onIndexModule(IndexModule indexModule, DiscoveryNode extensionNode) throws UnknownHostException {
        logger.info("onIndexModule index:" + indexModule.getIndex());
        final CompletableFuture<IndicesModuleResponse> inProgressFuture = new CompletableFuture<>();
        final CompletableFuture<AcknowledgedResponse> inProgressIndexNameFuture = new CompletableFuture<>();
        final TransportResponseHandler<AcknowledgedResponse> acknowledgedResponseHandler = new TransportResponseHandler<
            AcknowledgedResponse>() {
            @Override
            public void handleResponse(AcknowledgedResponse response) {
                logger.info("ACK Response" + response);
                inProgressIndexNameFuture.complete(response);
            }

            @Override
            public void handleException(TransportException exp) {
                inProgressIndexNameFuture.completeExceptionally(exp);
            }

            @Override
            public String executor() {
                return ThreadPool.Names.GENERIC;
            }

            @Override
            public AcknowledgedResponse read(StreamInput in) throws IOException {
                return new AcknowledgedResponse(in);
            }

        };

        final TransportResponseHandler<IndicesModuleResponse> indicesModuleResponseHandler = new TransportResponseHandler<
            IndicesModuleResponse>() {

            @Override
            public IndicesModuleResponse read(StreamInput in) throws IOException {
                return new IndicesModuleResponse(in);
            }

            @Override
            public void handleResponse(IndicesModuleResponse response) {
                logger.info("received {}", response);
                if (response.getIndexEventListener() == true) {
                    indexModule.addIndexEventListener(new IndexEventListener() {
                        @Override
                        public void beforeIndexRemoved(
                            IndexService indexService,
                            IndicesClusterStateService.AllocatedIndices.IndexRemovalReason reason
                        ) {
                            logger.info("Index Event Listener is called");
                            String indexName = indexService.index().getName();
                            logger.info("Index Name" + indexName.toString());
                            try {
                                logger.info("Sending extension request type: " + INDICES_EXTENSION_NAME_ACTION_NAME);
                                transportService.sendRequest(
                                    extensionNode,
                                    INDICES_EXTENSION_NAME_ACTION_NAME,
                                    new IndicesModuleRequest(indexModule),
                                    acknowledgedResponseHandler
                                );
                                
                                inProgressIndexNameFuture.whenComplete((r, e) -> {
                                    if (e == null) {
                                        inProgressFuture.complete(response);
                                    } else {
                                        inProgressFuture.completeExceptionally(e);
                                    }
                                });
                            } catch (Exception ex) {
                                inProgressFuture.completeExceptionally(ex);
                            }
                        }
                    });
                } else {
                    inProgressFuture.complete(response);
                }
            }

            @Override
            public void handleException(TransportException exp) {
                logger.error(new ParameterizedMessage("IndicesModuleRequest failed"), exp);
                inProgressFuture.completeExceptionally(exp);
            }

            @Override
            public String executor() {
                return ThreadPool.Names.GENERIC;
            }
        };

        try {
            logger.info("Sending extension request type: " + INDICES_EXTENSION_POINT_ACTION_NAME);
            transportService.sendRequest(
                extensionNode,
                INDICES_EXTENSION_POINT_ACTION_NAME,
                new IndicesModuleRequest(indexModule),
                indicesModuleResponseHandler
            );

            inProgressFuture
                .orTimeout(EXTENSION_REQUEST_WAIT_TIMEOUT, TimeUnit.SECONDS)
                .join();
            logger.info("Received response from Extension");
        } catch (final CompletionException e) {
            // ... exception handling logic here
        } catch (final CancellationException e) {
            // ... exception handling logic here
        }
    }
```

Reply:
I like this approach, essentially multi-threading the individual requests but ultimately waiting on all of them.